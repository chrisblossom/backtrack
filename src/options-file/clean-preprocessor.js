/* @flow */

import { toArray } from '../utils/object-utils';
import type { Clean, NormalizedClean } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value?: Clean | $ReadOnlyArray<Clean>,
};

function cleanPreprocessor({ value }: Args = {}): $ReadOnlyArray<
    NormalizedClean,
> {
    const allToArray = toArray(value);

    return allToArray.map((arg: Clean) => {
        return {
            del: toArray(arg.del),
            makeDirs: toArray(arg.makeDirs),
        };
    });
}

export { cleanPreprocessor };
