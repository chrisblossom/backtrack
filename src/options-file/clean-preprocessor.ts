import path from 'path';
import { toArray } from '../utils/object-utils';
import { buildPath, rootPath } from '../config/paths';
import { Clean, NormalizedClean } from '../types';

interface Args {
	value?: Clean | Clean[];
}

function cleanPreprocessor({ value }: Args = {}): NormalizedClean[] {
	const allToArray = toArray(value);

	return allToArray.map((arg: Clean) => {
		const makeDirs = toArray(arg.makeDirs).map((makeDir) => {
			const resolvedMakeDir = path.resolve(buildPath, makeDir);
			return resolvedMakeDir;
		});

		const copy = toArray(arg.copy).map((copyArgs) => {
			const resolvedSrc = path.resolve(rootPath, copyArgs.src);
			const resolvedDest = path.resolve(buildPath, copyArgs.dest);

			const result = {
				src: resolvedSrc,
				dest: resolvedDest,
				hash: copyArgs.hash,
			};

			if (result.hash === undefined) {
				delete result.hash;
			}

			return result;
		});

		return {
			del: toArray(arg.del),
			makeDirs,
			copy,
		};
	});
}

export { cleanPreprocessor };
