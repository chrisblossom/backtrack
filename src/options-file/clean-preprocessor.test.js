/* @flow */

const cleanPreprocessor = (args) =>
    require('./clean-preprocessor').cleanPreprocessor(args);

describe('cleanPreprocessor', () => {
    test('handles undefined', () => {
        const result = cleanPreprocessor();

        expect(result).toEqual([]);
    });

    test('converts all values to arrays and adds missing keys', () => {
        const value = [
            {
                del: '*',
            },
            {
                makeDirs: 'test',
            },
            {},
            {
                del: '*',
                makeDirs: 'test',
            },
        ];

        const result = cleanPreprocessor({ value });

        expect(result).toMatchSnapshot();
    });
});
