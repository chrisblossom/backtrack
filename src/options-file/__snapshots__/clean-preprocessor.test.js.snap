// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanPreprocessor converts all values to arrays and adds missing keys 1`] = `
Array [
  Object {
    "del": Array [
      "*",
    ],
    "makeDirs": Array [],
  },
  Object {
    "del": Array [],
    "makeDirs": Array [
      "test",
    ],
  },
  Object {
    "del": Array [],
    "makeDirs": Array [],
  },
  Object {
    "del": Array [
      "*",
    ],
    "makeDirs": Array [
      "test",
    ],
  },
]
`;
